Port CPU torch.geqrf to ATen

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56257 Fix MAGMA qr for empty batched inputs
* #56256 Add cuSOLVER path for torch.linalg.qr
* #56255 Remove size arguments for internal orgqr and geqrf calls
* #56254 Add non-allocating helper function for torch.linalg.qr
* #56253 Add cuBLAS path for batched torch.geqrf
* #56252 Add cuSOLVER path for torch.geqrf
* #56251 Port CUDA torch.geqrf to ATen
* #56250 Update internal code for torch.geqrf
* **#56249 Port CPU torch.geqrf to ATen**

This PR ports `torch.geqrf` from TH to ATen. CUDA path will be
implemented in a follow-up PR.
With ATen port support for complex and batched inputs is added.
There were no correctness tests, they are
added in this PR and I added OpInfo for this operation.

We can implement the QR decomposition as a composition of geqrf and
orgqr (torch.linalg.householder_product).
Also we can implement the least squares solver with geqrf + ormqr +
trtrs. So it's useful to have this function renewed at least for the
internal code.

Resolves https://github.com/pytorch/pytorch/issues/24705

Differential Revision: [D27907357](https://our.internmc.facebook.com/intern/diff/D27907357)

IvanYashchuk committed:
Port CPU torch.geqrf to ATen

This PR ports `torch.geqrf` from TH to ATen. CUDA path will be
implemented in a follow-up PR.
With ATen port support for complex and batched inputs is added.
There were no correctness tests, they are
added in this PR and I added OpInfo for this operation.

We can implement the QR decomposition as a composition of geqrf and
orgqr (torch.linalg.householder_product).
Also we can implement the least squares solver with geqrf + ormqr +
trtrs. So it's useful to have this function renewed at least for the
internal code.

Resolves https://github.com/pytorch/pytorch/issues/24705

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port CPU torch.geqrf to ATen"


This PR ports `torch.geqrf` from TH to ATen. CUDA path will be
implemented in a follow-up PR.
With ATen port support for complex and batched inputs is added.
There were no correctness tests, they are
added in this PR and I added OpInfo for this operation.

We can implement the QR decomposition as a composition of geqrf and
orgqr (torch.linalg.householder_product).
Also we can implement the least squares solver with geqrf + ormqr +
trtrs. So it's useful to have this function renewed at least for the
internal code.

Resolves https://github.com/pytorch/pytorch/issues/24705

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1606,17 +1606,38 @@ std::tuple<Tensor&, Tensor&> triangular_solve_out(const Tensor& self, const Tens
+static void apply_geqrf(const Tensor& self, const Tensor& tau, int64_t m, int64_t n) {
 #ifndef USE_LAPACK
-  AT_ERROR("qr: LAPACK library not found in compilation");
+  AT_ERROR("geqrf: LAPACK library not found in compilation");
```
TORCH_CHECK(false, ...)

mruberry replied:
We have a different error string for LAPACK not being available in the other linalg ops, right?

IvanYashchuk replied:
It's not consistent, some use the message as here and others use a nicer message. I'll change it here to a nicer message.

mruberry commented on test/test_linalg.py:
```diff
@@ -7420,13 +7420,40 @@ def test_nuclear_norm_out(self, device, dtype):
+    def test_geqrf(self, device, dtype):
+
+        def run_test(shape):
+            # NumPy outputs the result of geqrf operation
```
This comment is a little confusing as written. It's that NumPy doesn't have a function named geqrf, but np.linalg.qr with mode='raw' computes the same operation, so this test compares against that function.

mruberry commented on test/test_linalg.py:
```diff
@@ -7420,13 +7420,40 @@ def test_nuclear_norm_out(self, device, dtype):
+        batches = [(), (0, ), (2, ), (2, 1)]
+        ns = [5, 2, 0]
+        for batch, (m, n) in product(batches, product(ns, ns)):
+            # TODO: CUDA path doesn't work with batched or empty inputs
```
Assert this fails instead of just continuing so when the behavior changes the test demands an update, too

IvanYashchuk replied:
The test is updated in https://github.com/pytorch/pytorch/pull/56251. So can we leave this one as is? In other cases, I agree that it's better to assert failures and not skip the test.

IvanYashchuk replied:
I'll add the assert.

mruberry replied:
haha, really negotiating against yourself there

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3368,24 +3368,35 @@ def merge_dicts(*dicts):
-
-Rather, this directly calls the underlying LAPACK function `?geqrf`
+Computes a QR decomposition of :attr:`input`.
+Both `Q` and `R` matrices are stored in the same tensor `a`.
```
"in the same output tensor \`a\`."

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3368,24 +3368,35 @@ def merge_dicts(*dicts):
+The elements of `R` are stored on and above the diagonal.
+Elementary reflectors (or Householder vectors) implicitly defining matrix `Q`
+are stored below the diagonal.
+Result of this function can be used together with :func:`torch.linalg.householder_product`
```
"Result" -> "The results"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3368,24 +3368,35 @@ def merge_dicts(*dicts):
+are stored below the diagonal.
+Result of this function can be used together with :func:`torch.linalg.householder_product`
+to obtain the `Q` matrix or
+with :func:`torch.ormqr` that uses implicit representation of matrix `Q` for efficient matrix-matrix multiplication.
```
"with :func:\`torch.ormqr\` that uses implicit representation of matrix \`Q\` for efficient matrix-matrix multiplication." -> "with :func:\`torch.ormqr\`, which uses an implicit representation of the \`Q\` matrix, for an efficient matrix-matrix multiplication."

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3368,24 +3368,35 @@ def merge_dicts(*dicts):
+to obtain the `Q` matrix or
+with :func:`torch.ormqr` that uses implicit representation of matrix `Q` for efficient matrix-matrix multiplication.
+
+This function directly calls the underlying LAPACK function `geqrf`
```
This sentence seems redundant - maybe it can be removed?

IvanYashchuk replied:
Yes, we can remove it.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3368,24 +3368,35 @@ def merge_dicts(*dicts):
 This is a low-level function for calling LAPACK directly. This function
```
"for calling LAPACK's `geqrf` directly."  ?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3368,24 +3368,35 @@ def merge_dicts(*dicts):
+.. note::
+    To obtain explicit Q and R matrices it is recommended to use :func:`torch.linalg.qr`.
+
+.. note::
```
I would combine this note with the previous to make one "see also" note.

mruberry replied:
"See also :func:\`torch.linalg.qr\`, which computes explicit Q and R matrices, and :func:\`torch.linalg.lstsq\` with the \`\`driver="gels"\`\` option for a function that can solve matrix equations using a QR decomposition." ?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2017,6 +2017,22 @@ def sample_inputs_linalg_qr(op_info, device, dtype, requires_grad=False, **kwarg
 
+def sample_inputs_geqrf(op_info, device, dtype, requires_grad=False):
+    """
+    This function generates input for torch.geqrf
```
This comment doesn't hurt anything but I don't think it adds anything to the code, either

IvanYashchuk replied:
I agree, we can remove it.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2017,6 +2017,22 @@ def sample_inputs_linalg_qr(op_info, device, dtype, requires_grad=False, **kwarg
+    out = []
+    for batch, (m, n) in product(batches, product(ns, ns)):
+        # TODO: CUDA path doesn't work with batched or empty inputs
+        if 'cuda' in device and (batch != () or m == 0 or n == 0):
```
```
if torch.device(device).type == 'cuda'
```

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2017,6 +2017,22 @@ def sample_inputs_linalg_qr(op_info, device, dtype, requires_grad=False, **kwarg
+    """
+    batches = [(), (0, ), (2, ), (1, 1)]
+    ns = [5, 2, 0]
+    out = []
```
"out" -> "samples"

mruberry reviewed (commented):
Overall this looks great, as usual; I've suggested a few tweaks and also pinged @ngimel since she's been reviewing most TH->ATen ports.

IvanYashchuk committed:
Update on "Port CPU torch.geqrf to ATen"


This PR ports `torch.geqrf` from TH to ATen. CUDA path will be
implemented in a follow-up PR.
With ATen port support for complex and batched inputs is added.
There were no correctness tests, they are
added in this PR and I added OpInfo for this operation.

We can implement the QR decomposition as a composition of geqrf and
orgqr (torch.linalg.householder_product).
Also we can implement the least squares solver with geqrf + ormqr +
trtrs. So it's useful to have this function renewed at least for the
internal code.

Resolves https://github.com/pytorch/pytorch/issues/24705

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port CPU torch.geqrf to ATen"


This PR ports `torch.geqrf` from TH to ATen. CUDA path will be
implemented in a follow-up PR.
With ATen port support for complex and batched inputs is added.
There were no correctness tests, they are
added in this PR and I added OpInfo for this operation.

We can implement the QR decomposition as a composition of geqrf and
orgqr (torch.linalg.householder_product).
Also we can implement the least squares solver with geqrf + ormqr +
trtrs. So it's useful to have this function renewed at least for the
internal code.

Resolves https://github.com/pytorch/pytorch/issues/24705

[ghstack-poisoned]

IvanYashchuk commented:
@mruberry thank you for your suggestions! I've updated this pull request.

IvanYashchuk committed:
Update on "Port CPU torch.geqrf to ATen"


This PR ports `torch.geqrf` from TH to ATen. CUDA path will be
implemented in a follow-up PR.
With ATen port support for complex and batched inputs is added.
There were no correctness tests, they are
added in this PR and I added OpInfo for this operation.

We can implement the QR decomposition as a composition of geqrf and
orgqr (torch.linalg.householder_product).
Also we can implement the least squares solver with geqrf + ormqr +
trtrs. So it's useful to have this function renewed at least for the
internal code.

Resolves https://github.com/pytorch/pytorch/issues/24705

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port CPU torch.geqrf to ATen"


This PR ports `torch.geqrf` from TH to ATen. CUDA path will be
implemented in a follow-up PR.
With ATen port support for complex and batched inputs is added.
There were no correctness tests, they are
added in this PR and I added OpInfo for this operation.

We can implement the QR decomposition as a composition of geqrf and
orgqr (torch.linalg.householder_product).
Also we can implement the least squares solver with geqrf + ormqr +
trtrs. So it's useful to have this function renewed at least for the
internal code.

Resolves https://github.com/pytorch/pytorch/issues/24705

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1642,6 +1669,102 @@ static void apply_geqrf(Tensor& self, Tensor& tau, int64_t m, int64_t n) {
 }
 
+static void geqrf_out_helper(const Tensor& input, const Tensor& QR, const Tensor& tau) {
+  TORCH_INTERNAL_ASSERT(input.dim() >= 2);
```
Do you want to make the TORCH_INTERNAL_ASSERT_DEBUG_ONLY? You've already checked all these conditions, and this function is not user facing.

mruberry commented:
@IvanYashchuk just ping me when you're happy with this PR, @IvanYashchuk, and I'll start merging this stack
